redhat_subscription: do not use defaults from the conf file

mkrizek opened the pull request:
##### SUMMARY
As noted by @kahowell in https://github.com/ansible/ansible/issues/25664#issuecomment-308442955, we seem to be better off not taking defaults from `/etc/rhsm/rhsm.conf` and only pass params to subscription-manager that were explicitly passed to the module.

##### ISSUE TYPE

 - Bugfix Pull Request

##### COMPONENT NAME
redhat_subscription

##### ANSIBLE VERSION
<!--- Paste verbatim output from "ansible --version" between quotes below -->
```
2.5
```


##### ADDITIONAL INFORMATION

mkrizek committed:
redhat_subscription: do not use defaults from the conf file

alikins commented:
For the state=present case that requires a register, it looks like 'subscription-manager config' is called with the module args, the 'subscription-manager register' is called with some subset of the module options as cli args.

Haven't tested, but it seems like the `rhsm.configure(**module.params)` will update rhsm.conf with the module defaults. The module defaults could be different? 

Most of the options are also provided as 'subman register' cli args as well, so maybe that resets them.

First thought is that rhsm.configure should only change values that were specified by task (and/or, count those difference in check mode).

Could the cli/config that don't have subman cli args (if any) be a problem?

kahowell commented:
> For the state=present case that requires a register, it looks like 'subscription-manager config' is called with the module args, the 'subscription-manager register' is called with some subset of the module options as cli args.

I looked through subscription-manager code, and it looks like we don't always save proxy credentials, so it is probably still worth calling `subscription-manager config`.

> Haven't tested, but it seems like the rhsm.configure(**module.params) will update rhsm.conf with the module defaults. The module defaults could be different?

Looks like that happens. I see `['/sbin/subscription-manager', 'config', '--server.hostname=None', '--server.proxy_hostname=None', '--server.proxy_user=None', '--server.insecure=None', '--server.proxy_password=None', '--rhsm.baseurl=None', '--server.proxy_port=None']` being invoked.

> First thought is that rhsm.configure should only change values that were specified by task (and/or, count those difference in check mode).

Yeah, I think this could be as simple as checking `is not None` in `configure()`...

> Could the cli/config that don't have subman cli args (if any) be a problem?

We've added a few options, an up-to-date list can be seen here: https://github.com/candlepin/subscription-manager/blob/master/etc-conf/rhsm.conf or via `subscription-manager config`. IMO the ansible module does not need to support all of the config options. (User can always template `/etc/rhsm/rhsm.conf` instead).

Food for thought: another option would be to allow the user to (attempt) to set any config key via `params` as is done in the `yum_repository` module. This would keep this module from needing changes whenever subscription-manager develops a new configuration option. (Though for KISS reasons, I'd actually prefer removing all options that can be specified by config, and directing users to template `/etc/rhsm/rhsm.conf` if necessary...) @alikins @mkrizek thoughts?

mkrizek committed:
Only pass options specified by user to the config

mkrizek commented:
> Looks like that happens. I see ['/sbin/subscription-manager', 'config', '--server.hostname=None', '--server.proxy_hostname=None', '--server.proxy_user=None', '--server.insecure=None', '--server.proxy_password=None', '--rhsm.baseurl=None', '--server.proxy_port=None'] being invoked.
> 
> > First thought is that rhsm.configure should only change values that were specified by task (and/or, count those difference in check mode).
> 
> Yeah, I think this could be as simple as checking is not None in configure()...

Good catch, fixed.

> 
> > Could the cli/config that don't have subman cli args (if any) be a problem?
> 
> We've added a few options, an up-to-date list can be seen here: https://github.com/candlepin/subscription-manager/blob/master/etc-conf/rhsm.conf or via subscription-manager config. IMO the ansible module does not need to support all of the config options. (User can always template /etc/rhsm/rhsm.conf instead).
> 
> Food for thought: another option would be to allow the user to (attempt) to set any config key via params as is done in the yum_repository module. This would keep this module from needing changes whenever subscription-manager develops a new configuration option. (Though for KISS reasons, I'd actually prefer removing all options that can be specified by config, and directing users to template /etc/rhsm/rhsm.conf if necessary...) @alikins @mkrizek thoughts?

@kahowell Actually, the `params` option was removed from the `yum_repository` module mostly because of security reasons and is discouraged to add to other modules.

@kahowell @alikins anything else we can do here before merging?

alikins commented:
@kahowell @mkrizek Only concern would be that if it could break existing playbooks if they rely on defaults from rhsm.conf

mkrizek commented:
> @kahowell @mkrizek Only concern would be that if it could break existing playbooks if they rely on defaults from rhsm.conf

@alikins wouldn't `subscription-manager` use defaults from `rhsm.conf` even if the module doesn't pass them explicitly?

alikins commented:
@mkrizek "wouldn't subscription-manager use defaults from rhsm.conf even if the module doesn't pass them explicitly"  .Looking at the code again, I would agree that subman will use the rhsm.conf values if the module doesn't specify.

It looks like the pr cleans up any places where it could try to provide a redhat_subscription.py specific default to use, which seems like the correct thing to me. So I think the patch looks good.

If the patch avoids having:
- a) the updates to rhsm.conf via 'subman configure' 
- b) the args passed to Rhsm.register() 

use values/defaults that are not None, or were not provided by the playbook/task (ie, defaults fabricated by redhat_subscription.py) that should work.  

Afaict, the patch avoids those issues so looks good to me.
